SCAutolib/models/card.py

```diff
@@ -0,0 +1,229 @@
+        """
+
+        self.user = user
+        assert self.user.card_dir.exists(), "Card root directory doesn't exists"
+
+        self._private_key = key
+        self._cert = cert
+
+        self._service_name = f"virt_cacard_{self.user.username}"
+        self._service_location = Path(
+            f"/etc/systemd/system/{self._service_name}.service")
+        self._insert = insert
+        self._nssdb = self.user.card_dir.joinpath("db")
+        self._softhsm2_conf = softhsm2_conf
+
+    def __enter__(self):
+        """
+        Start of context manger for virtual smart card. The card would be
+        inserted if ``insert`` parameter in constructor is specified.
+
+        :return: self
+        """
+        assert self._service_location.exists(), \
+            "Service for virtual sc doesn't exists."
+        if self._insert:
+            self.insert()
+        return self
+
+    def __exit__(self, exp_type, exp_value, exp_traceback):
+        """
+        End of context manager for virtual smart card. If any exception was
+        raised in the current context, it would be logged as an error.
+
+        :param exp_type: Type of the exception
+        :param exp_value: Value for the exception
+        :param exp_traceback: Traceback of the exception
+        """
+        if exp_type is not None:
+            logger.error("Exception in virtual smart card context")
+            logger.error(format_exc())
+        self.remove()
+
+    @property
+    def softhsm2_conf(self):
+        return self._softhsm2_conf
+
+    @softhsm2_conf.setter
+    def softhsm2_conf(self, conf: SoftHSM2Conf):
+        assert conf.conf_path.exists(), "File doesn't exist"
+        self._softhsm2_conf = conf.conf_path
+
+    def insert(self):
+        """
+        Insert virtual smart card by starting the corresponding service.
+        """
+        cmd = ["systemctl", "start", self._service_name]
+        out = run(cmd, check=True)
+        time.sleep(2)  # to prevent error with fast restarting of the service
+        logger.info(f"Smart card {self._service_name} is inserted")
+        return out
+
+    def remove(self):
+        """
+        Remove the virtual card by stopping the service
+        """
+        cmd = ["systemctl", "stop", self._service_name]
+        out = run(cmd, check=True)
+        time.sleep(2)  # to prevent error with fast restarting of the service
+        logger.info(f"Smart card {self._service_name} is removed")
+        return out
+
+    def enroll(self):
+        """
+        Upload certificate and private key to the virtual smart card (upload to
+        NSS database) with pkcs11-tool.
+        """
+        cmd = ["pkcs11-tool", "--module", "libsofthsm2.so", "--slot-index",
+               '0', "-w", self._private_key, "-y", "privkey", "--label",
+               f"'{self.user.username}'", "-p", self.user.pin, "--set-id", "0",
+               "-d", "0"]
+        run(cmd, env={"SOFTHSM2_CONF": self._softhsm2_conf})
```

Comment:
You don't check for exit code

Reply:
will set check for the return code as a default behaviour for `run` function